app/services/tool_filter_service.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+
+class ToolFilterService
+  def initialize(params)
+    @params = params
+  end
+
+  def call
+    tool_groups = []
+    ToolGroup.all.each { |tool_group| tool_groups << tool_group if match_params(tool_group, @params) }
+    grouped_array = group_resources(tool_groups)
+    order_results(grouped_array)
+  end
+
+  private
+
+  def order_results(grouped_array)
+    result = grouped_array.map do |key, values|
+      counter = values.size
+      sum = values.sum { |o| o[:tool_group_suggestions_weight] * o[:resource_tool_group_suggestions_weight] }
+      average = sum / counter
+      [Resource.find(key), average]
+    end
+
+    sorted_result = result.sort_by { |subarray| -subarray[1] }
+    sorted_result.map { |subarray| subarray[0] }
+  end
+
+  def group_resources(tool_groups)
+    result = tool_groups.flat_map do |tool_group|
+      tool_group.resource_tool_groups.map do |resource|
+        {
+          tool_group_suggestions_weight: tool_group.suggestions_weight,
+          resource_id: resource.resource_id,
+          resource_tool_group_suggestions_weight: resource.suggestions_weight
+        }
+      end
+    end
+
+    result.group_by { |hash| hash[:resource_id] }
+  end
+
+  def match_params(tool_group, params)
+    if params.has_key?(:filter)
+      country = params["filter"]["country"]&.upcase if params["filter"]["country"].present?
+      languages = params["filter"]["language"] if params["filter"]["language"].present?
+      openness = params["filter"]["openness"].to_i if params["filter"]["openness"].present?
+      confidence = params["filter"]["confidence"].to_i if params["filter"]["confidence"].present?
+    end
+
+    if country.nil? & languages.nil? & openness.nil? & confidence.nil?
+      if no_rules_for(tool_group)
+        return true
+      else
+        return false
+      end
+    end
+
+    return true if no_rules_for(tool_group)
+
+    # Rule Countries
+    if tool_group.rule_countries.any?
+      if country
+        if tool_group.rule_countries.any? { |o| o.countries.exclude?(country) && o.negative_rule }
+          return true
```

Comment:
a tool-group defined with the following 2 rules:
- languages: de
- countries: FR

when issuing the following query the group incorrectly matches:
`https://mobile-content-api-stage.cru.org/resources/suggestions?filter[country]=FR`

The group shouldn't match because the languages rule was not satisfied by the provided filters

Reply:
I think this initial `if` statement can go away and just the 2 `elsifs` are needed

Reply:
Both rules are `negative_rule: false ` ?

Reply:
It should be solved now